ccl/kvccl/kvtenantccl: TestTenantUpgradeFailure failed

ccl/kvccl/kvtenantccl.TestTenantUpgradeFailure [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3446147&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3446147&tab=artifacts#/) on master @ [2baefd353239ddc4e7f4299b8930001675b99aa9](https://github.com/cockroachdb/cockroach/commits/2baefd353239ddc4e7f4299b8930001675b99aa9):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2fbf0cd]

```
Stack: 

```
goroutine 15676 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0026a7830, 0x935b050, 0xc0023be938)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x5dccd20, 0xb2c77d0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).started.func1(0xc0003eb1e0, 0x99f835536048002, 0xc003680310, 0x7, 0xc002900a50, 0xc0039bc600, 0x0, 0xc00166c270, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:267 +0x2cd
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).update.func1(0x933f7f8, 0xc0029e6b40, 0xc0003eb1e0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:218 +0x9bd
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x933f7f8, 0xc0029e6b40, 0xc0003eb1e0, 0xc0012f7400, 0x272d6e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:818 +0x5f
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc0003eb1e0, 0x933f7f8, 0xc0029e6b40, 0xc007009260, 0xc0003eb1e0, 0x100000015)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:890 +0xef
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc003b79380, 0x933f7f8, 0xc0029e6b40, 0xc001285348, 0xc001606000, 0x12)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:817 +0x125
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0xc00677c780, 0x933f7f8, 0xc0029e6b40, 0x0, 0xc001285348, 0xc0029e6bd0, 0xc0029e6bd0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:728 +0xf5
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).update(0xc00677c780, 0x933f7f8, 0xc0029e6b40, 0x0, 0x933f700, 0xc0012853e0, 0x5dca3e0, 0xc0a98b8)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:147 +0x1b9
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:138
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).started(0xc00677c780, 0x933f7f8, 0xc0029e6b40, 0x0, 0x5cab640, 0xc0018e8e70)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:257 +0x9e
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc00430a000, 0x933f7f8, 0xc0029e6b40, 0x6138720, 0xc003c38720, 0x93000a8, 0xc003cb0c50, 0xc00677c780, 0x63d5c24, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1172 +0xaa5
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc00430a000, 0x933f750, 0xc00354af80, 0x93000a8, 0xc003cb0c50, 0xc00677c780, 0x63d5c24, 0x7, 0xc0036885b8, 0x16, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:415 +0x61d
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2(0x935b050, 0xc0023be938)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:889 +0x235
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc0026a7830, 0x935b050, 0xc0023be938, 0xc0023be780, 0x0, 0xc0056f5320)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x162
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x28e
```
<details><summary>Log preceding fatal error</summary>
<p>

```
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227  
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +__level_____count____size___score______in__ingest(sz_cnt)____move(sz_cnt)___write(sz_cnt)____read___r-amp___w-amp
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +    WAL         4   1.5 M       -   1.4 M       -       -       -       -   1.5 M       -       -       -     1.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +      0         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +      1         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +      2         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +      3         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +      4         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +      5         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +      6         0     0 B       -     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +  total         0     0 B       -   1.5 M     0 B       0     0 B       0   1.5 M       0     0 B       0     1.0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +  flush         0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +compact         0     0 B             0 B          (size == estimated-debt, in = in-progress-bytes)
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +  ctype         0       0       0       0       0  (default, delete, elision, move, read)
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 + memtbl         4   3.8 M
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +zmemtbl         0     0 B
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 +   ztbl         0     0 B
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 + bcache         0     0 B    0.0%  (score == hit-rate)
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 + tcache         0     0 B    0.0%  (score == hit-rate)
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 + titers         0
I210915 05:41:50.237910 8895 kv/kvserver/store.go:2700  [n1,s1] 1227 + filter         -       -    0.0%  (score == utility)
I210915 05:41:56.915965 13923 2@util/log/event_log.go:32  [-] 1228 ={"Timestamp":1631684516915943938,"EventType":"runtime_stats","MemRSSBytes":897806336,"GoroutineCount":473,"MemStackSysBytes":4653056,"GoAllocBytes":44401744,"GoTotalBytes":124358680,"GoStatsStaleness":0.001699358,"HeapFragmentBytes":18946992,"HeapReservedBytes":43425792,"HeapReleasedBytes":22790144,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.29999265,"CPUUserPercent":67.79834,"CPUSysPercent":2.6999338,"GCPausePercent":0.005375258,"GCRunCount":88,"NetHostRecvBytes":69882,"NetHostSendBytes":69882}
I210915 05:41:57.367530 14177 2@util/log/event_log.go:32  [-] 1229 ={"Timestamp":1631684517367504766,"EventType":"runtime_stats","MemRSSBytes":895016960,"GoroutineCount":473,"MemStackSysBytes":5767168,"GoAllocBytes":52701344,"GoTotalBytes":124358680,"GoStatsStaleness":0.002511061,"HeapFragmentBytes":17930080,"HeapReservedBytes":35028992,"HeapReleasedBytes":22790144,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.29996294,"CPUUserPercent":65.991844,"CPUSysPercent":2.8996415,"GCPausePercent":0.0053747254,"GCRunCount":88,"NetHostRecvBytes":71106,"NetHostSendBytes":71106}
I210915 05:42:00.236557 8965 2@util/log/event_log.go:32  [n1] 1230 ={"Timestamp":1631684520236535956,"EventType":"runtime_stats","MemRSSBytes":901275648,"GoroutineCount":473,"MemStackSysBytes":5308416,"GoAllocBytes":62085888,"GoTotalBytes":124383256,"GoStatsStaleness":0.001784147,"HeapFragmentBytes":17515776,"HeapReservedBytes":26542080,"HeapReleasedBytes":22765568,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.29997146,"CPUUserPercent":64.393875,"CPUSysPercent":3.3996766,"GCPausePercent":0.0043919217,"GCRunCount":89,"NetHostRecvBytes":72734,"NetHostSendBytes":72734}
I210915 05:42:03.833633 9878 server/auto_upgrade.go:44  [n1] 1231  auto upgrade disabled by testing
I210915 05:42:06.917439 13923 2@util/log/event_log.go:32  [-] 1232 ={"Timestamp":1631684526917415039,"EventType":"runtime_stats","MemRSSBytes":909758464,"GoroutineCount":474,"MemStackSysBytes":5603328,"GoAllocBytes":66501336,"GoTotalBytes":127373336,"GoStatsStaleness":0.002412997,"HeapFragmentBytes":16975144,"HeapReservedBytes":25362432,"HeapReleasedBytes":19775488,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.299956,"CPUUserPercent":64.89048,"CPUSysPercent":4.1993837,"GCPausePercent":0.0056797564,"GCRunCount":92,"NetHostRecvBytes":73078,"NetHostSendBytes":73078}
I210915 05:42:07.367385 14177 2@util/log/event_log.go:32  [-] 1233 ={"Timestamp":1631684527367365634,"EventType":"runtime_stats","MemRSSBytes":906911744,"GoroutineCount":475,"MemStackSysBytes":5865472,"GoAllocBytes":74236576,"GoTotalBytes":127373336,"GoStatsStaleness":0.001637828,"HeapFragmentBytes":16235872,"HeapReservedBytes":18104320,"HeapReleasedBytes":19775488,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.30000412,"CPUUserPercent":65.00089,"CPUSysPercent":4.2000575,"GCPausePercent":0.0056806677,"GCRunCount":92,"NetHostRecvBytes":74519,"NetHostSendBytes":74519}
I210915 05:42:10.237635 8965 2@util/log/event_log.go:32  [n1] 1234 ={"Timestamp":1631684530237610408,"EventType":"runtime_stats","MemRSSBytes":907845632,"GoroutineCount":477,"MemStackSysBytes":5865472,"GoAllocBytes":83041416,"GoTotalBytes":128184344,"GoStatsStaleness":0.00225412,"HeapFragmentBytes":15713144,"HeapReservedBytes":10633216,"HeapReleasedBytes":18964480,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.29996783,"CPUUserPercent":65.09302,"CPUSysPercent":3.7995923,"GCPausePercent":0.005739374,"GCRunCount":93,"NetHostRecvBytes":72744,"NetHostSendBytes":72744}
I210915 05:42:16.917221 13923 2@util/log/event_log.go:32  [-] 1235 ={"Timestamp":1631684536917192788,"EventType":"runtime_stats","MemRSSBytes":917848064,"GoroutineCount":476,"MemStackSysBytes":5636096,"GoAllocBytes":83315528,"GoTotalBytes":128987160,"GoStatsStaleness":0.001755924,"HeapFragmentBytes":15963320,"HeapReservedBytes":11141120,"HeapReleasedBytes":18161664,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.3000067,"CPUUserPercent":62.701397,"CPUSysPercent":4.2000937,"GCPausePercent":0.004575182,"GCRunCount":96,"NetHostRecvBytes":74109,"NetHostSendBytes":74109}
I210915 05:42:17.367958 14177 2@util/log/event_log.go:32  [-] 1236 ={"Timestamp":1631684537367935641,"EventType":"runtime_stats","MemRSSBytes":918745088,"GoroutineCount":476,"MemStackSysBytes":4653056,"GoAllocBytes":52165320,"GoTotalBytes":129585176,"GoStatsStaleness":0.001714521,"HeapFragmentBytes":17827128,"HeapReservedBytes":42008576,"HeapReleasedBytes":17563648,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.29998288,"CPUUserPercent":63.89635,"CPUSysPercent":4.1997604,"GCPausePercent":0.005542514,"GCRunCount":97,"NetHostRecvBytes":70547,"NetHostSendBytes":70547}
I210915 05:42:20.059418 15365 sql/sqlliveness/slstorage/slstorage.go:305  [intExec=expire-sessions] 1237  deleted session f8ee354435e24d86b9a7bb08e3d5df19 which expired at 1631684536.505843389,0
I210915 05:42:20.237673 8965 2@util/log/event_log.go:32  [n1] 1238 ={"Timestamp":1631684540237651996,"EventType":"runtime_stats","MemRSSBytes":910688256,"GoroutineCount":476,"MemStackSysBytes":5636096,"GoAllocBytes":58111168,"GoTotalBytes":129937432,"GoStatsStaleness":0.001723723,"HeapFragmentBytes":16812864,"HeapReservedBytes":36446208,"HeapReleasedBytes":17211392,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.2999987,"CPUUserPercent":61.19973,"CPUSysPercent":4.399981,"GCPausePercent":0.0055706054,"GCRunCount":98,"NetHostRecvBytes":75457,"NetHostSendBytes":75457}
I210915 05:42:26.917333 13923 2@util/log/event_log.go:32  [-] 1239 ={"Timestamp":1631684546917313060,"EventType":"runtime_stats","MemRSSBytes":932937728,"GoroutineCount":476,"MemStackSysBytes":4653056,"GoAllocBytes":53418768,"GoTotalBytes":133328920,"GoStatsStaleness":0.001734401,"HeapFragmentBytes":17614064,"HeapReservedBytes":44711936,"HeapReleasedBytes":13819904,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.29999632,"CPUUserPercent":60.499256,"CPUSysPercent":4.0999494,"GCPausePercent":0.0056951,"GCRunCount":101,"NetHostRecvBytes":74101,"NetHostSendBytes":74101}
I210915 05:42:27.368714 14177 2@util/log/event_log.go:32  [-] 1240 ={"Timestamp":1631684547368692699,"EventType":"runtime_stats","MemRSSBytes":933273600,"GoroutineCount":476,"MemStackSysBytes":5046272,"GoAllocBytes":59599736,"GoTotalBytes":133328920,"GoStatsStaleness":0.002246072,"HeapFragmentBytes":16807048,"HeapReservedBytes":38944768,"HeapReleasedBytes":13819904,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.2999774,"CPUUserPercent":59.195538,"CPUSysPercent":3.9996986,"GCPausePercent":0.0047270632,"GCRunCount":101,"NetHostRecvBytes":75911,"NetHostSendBytes":75911}
I210915 05:42:30.232237 8896 2@gossip/gossip.go:567  [n1] 1241  gossip status (ok, 1 node)
I210915 05:42:30.232237 8896 2@gossip/gossip.go:567  [n1] 1241 +gossip client (0/3 cur/max conns)
I210915 05:42:30.232237 8896 2@gossip/gossip.go:567  [n1] 1241 +gossip server (0/3 cur/max conns, infos 0/0 sent/received, bytes 0B/0B sent/received)
I210915 05:42:30.232237 8896 2@gossip/gossip.go:567  [n1] 1241 +gossip connectivity
I210915 05:42:30.232237 8896 2@gossip/gossip.go:567  [n1] 1241 +  n1 [sentinel];
I210915 05:42:30.239336 8965 2@util/log/event_log.go:32  [n1] 1242 ={"Timestamp":1631684550239310505,"EventType":"runtime_stats","MemRSSBytes":926257152,"GoroutineCount":476,"MemStackSysBytes":5701632,"GoAllocBytes":65823584,"GoTotalBytes":131067928,"GoStatsStaleness":0.002561301,"HeapFragmentBytes":16555168,"HeapReservedBytes":30056448,"HeapReleasedBytes":16080896,"CGoAllocBytes":5938624,"CGoTotalBytes":10354688,"CGoCallRate":0.29995033,"CPUUserPercent":60.489983,"CPUSysPercent":3.8993542,"GCPausePercent":0.005410624,"GCRunCount":102,"NetHostRecvBytes":68926,"NetHostSendBytes":68926}
I210915 05:42:31.411250 8872 3@vendor/github.com/cockroachdb/pebble/event.go:631  [n1,pebble,s1] 1243  [JOB 6] WAL created 000007
I210915 05:42:36.029993 9878 server/auto_upgrade.go:44  [n1] 1244  auto upgrade disabled by testing
I210915 05:42:36.385335 13999 jobs/adopt.go:106  [-] 1245  claimed 1 jobs
I210915 05:42:36.404107 15697 jobs/adopt.go:247  [-] 1246  job 693417071789047809: resuming execution
I210915 05:42:36.420455 15654 jobs/registry.go:1150  [sql] 1247  MIGRATION job 693417071789047809: stepping through state running with error: <nil>
I210915 05:42:36.481645 15654 jobs/registry.go:1150  [sql] 1248  MIGRATION job 693417071789047809: stepping through state succeeded with error: <nil>
I210915 05:42:36.918016 13923 2@util/log/event_log.go:32  [-] 1249 ={"Timestamp":1631684556917997712,"EventType":"runtime_stats","MemRSSBytes":926072832,"GoroutineCount":475,"MemStackSysBytes":5242880,"GoAllocBytes":75041024,"GoTotalBytes":131837976,"GoStatsStaleness":0.001801746,"HeapFragmentBytes":15243008,"HeapReservedBytes":23379968,"HeapReleasedBytes":15310848,"CGoAllocBytes":10133056,"CGoTotalBytes":14553088,"CGoCallRate":0.39997256,"CPUUserPercent":63.995613,"CPUSysPercent":4.6996775,"GCPausePercent":0.0064502377,"GCRunCount":105,"NetHostRecvBytes":94077,"NetHostSendBytes":94077}
I210915 05:42:37.055308 15676 jobs/registry.go:1150  [sql] 1250  AUTO CREATE STATS job 693417269640069122: stepping through state running with error: <nil>
I210915 05:42:37.059992 14274 server/settingswatcher/settings_watcher.go:112  [rangefeed=settings] 1251  set cluster version to: 21.1-1162
I210915 05:42:37.061162 14001 server/settingswatcher/settings_watcher.go:112  [rangefeed=settings] 1252  set cluster version to: 21.1-1162
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantUpgradeFailure PKG=./pkg/ccl/kvccl/kvtenantccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/kv fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgradeFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/jobs/helpers_test.go
+++ b/pkg/jobs/helpers_test.go
@@ -71,6 +71,12 @@ func (j *Job) Started(ctx context.Context) error {
 	return j.started(ctx, nil /* txn */)
 }
 
+// Reverted is a wrapper around the internal function that moves a job to the
+// reverting state.
+func (j *Job) Reverted(ctx context.Context, err error) error {
+	return j.reverted(ctx, nil /* txn */, err, nil)
+}
+
 // Paused is a wrapper around the internal function that moves a job to the
 // paused state.
 func (j *Job) Paused(ctx context.Context) error {

--- a/pkg/jobs/jobs.go
+++ b/pkg/jobs/jobs.go
@@ -16,7 +16,6 @@ import (
 	"reflect"
 	"sync/atomic"
 
-	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/security"
@@ -263,7 +262,14 @@ func (j *Job) started(ctx context.Context, txn *kv.Txn) error {
 			md.Payload.StartedMicros = timeutil.ToUnixMicros(j.registry.clock.Now().GoTime())
 			ju.UpdatePayload(md.Payload)
 		}
-		if j.registry.settings.Version.IsActive(ctx, clusterversion.RetryJobsWithExponentialBackoff) {
+		// md.RunStats can be nil because of the timing of version-update when exponential-backoff
+		// gets activated. It may happen that backoff is not activated when Update() function was
+		// called, which will cause to not populate md.RunStats. However, when the code reaches this
+		// point, version update may have been updated to enable backoff. In this case, we can skip
+		// updating num_runs and last_run, treating this job run as if backoff was not activated.
+		//
+		// TODO (sajjad): Update this comment after version 22.2 has been released.
+		if md.RunStats != nil {
 			ju.UpdateRunStats(md.RunStats.NumRuns+1, j.registry.clock.Now().GoTime())
 		}
 		return nil
@@ -541,7 +547,12 @@ func (j *Job) reverted(
 			}
 			ju.UpdateStatus(StatusReverting)
 		}
-		if j.registry.settings.Version.IsActive(ctx, clusterversion.RetryJobsWithExponentialBackoff) {
+		// md.RunStats will be nil if clusterversion.RetryJobsWithExponentialBackoff
+		// was not active when Update was called above. In this case, we skip updating
+		// the runStats, treating this job run as if backoff is not active.
+		//
+		// TODO (sajjad): Update this comment after version 22.2 has been released.
+		if md.RunStats != nil {
 			// We can reach here due to a failure or due to the job being canceled.
 			// We should reset the exponential backoff parameters if the job was not
 			// canceled. Note that md.Status will be StatusReverting if the job

--- a/pkg/jobs/registry_external_test.go
+++ b/pkg/jobs/registry_external_test.go
@@ -24,10 +24,12 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql"
@@ -39,6 +41,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
@@ -723,4 +726,106 @@ SELECT unnest(execution_errors)
 		close(seventhRun.resume)
 		require.Regexp(t, err3, registry.WaitForJobs(ctx, ie, []jobspb.JobID{id}))
 	})
+}
+
+// TestExponentialBackoffUpgradeRacesWithUpdate tests that the system does not
+// encounter a nil pointer panic if the version is activated during an update.
+// When exponential-backoff is not active, num_runs and last_run columns are not
+// present in system.jobs table. As we checked version updates at multiple
+// locations and query system.jobs table during a job lifecycle, this test
+// checks that version update between two cluster-version checks does not result
+// in an error.
+func TestExponentialBackoffUpgradeRacesWithUpdate(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	var stmtFilter atomic.Value
+	resetStmtFilter := func() { stmtFilter.Store(func(context.Context, string) {}) }
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			Knobs: base.TestingKnobs{
+				Server: &server.TestingKnobs{
+					DisableAutomaticVersionUpgrade: 1,
+					BinaryVersionOverride: clusterversion.ByKey(
+						clusterversion.RetryJobsWithExponentialBackoff - 1),
+				},
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+				SQLExecutor: &sql.ExecutorTestingKnobs{
+					StatementFilter: func(ctx context.Context, data *sessiondata.SessionData, s string, err error) {
+						stmtFilter.Load().(func(context.Context, string))(ctx, s)
+					},
+				},
+			},
+		},
+	}
+
+	jobs.RegisterConstructor(jobspb.TypeImport, func(_ *jobs.Job, cs *cluster.Settings) jobs.Resumer {
+		return jobs.FakeResumer{}
+	})
+	defer jobs.ResetConstructors()
+	type updateFunc = func(j *jobs.Job, ctx context.Context) error
+	// Sketch of the test:
+	//  * Create the cluster in the old version.
+	//  * Create the job but don't start it.
+	//  * Install a hook to block the select of the update.
+	//  * Launch a goroutine to run the relevant update.
+	//  * Wait for the update to block and unregister the hook.
+	//  * Upgrade the version.
+	//  * Unblock the update, see no panic.
+	run := func(t *testing.T, f updateFunc) {
+		resetStmtFilter()
+		ctx := context.Background()
+		tc := testcluster.StartTestCluster(t, 1, clusterArgs)
+		defer tc.Stopper().Stop(ctx)
+		s := tc.Server(0)
+
+		sqlDB := tc.ServerConn(0)
+		tdb := sqlutils.MakeSQLRunner(sqlDB)
+		registry := s.JobRegistry().(*jobs.Registry)
+		var sj *jobs.StartableJob
+		require.NoError(t, s.DB().Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			return registry.CreateStartableJobWithTxn(ctx, &sj, registry.MakeJobID(), txn, jobs.Record{
+				Details:  jobspb.ImportDetails{},
+				Progress: jobspb.ImportProgress{},
+			})
+		}))
+		blocked := make(chan chan struct{})
+
+		stmtFilter.Store(func(ctx context.Context, s string) {
+			if !strings.HasPrefix(s, `SELECT status, payload, progress`) {
+				return
+			}
+			resetStmtFilter() // so nobody else gets blocked
+			unblock := make(chan struct{})
+			blocked <- unblock
+			<-unblock
+		})
+		updateErr := make(chan error)
+		go func() { updateErr <- f(sj.Job, ctx) }()
+		unblock := <-blocked
+		resetStmtFilter() // so we don't catch other jobs
+		tdb.Exec(t,
+			`SET CLUSTER SETTING version = $1`, clusterversion.ByKey(clusterversion.RetryJobsWithExponentialBackoff).String(),
+		)
+		select {
+		case <-unblock:
+			t.Fatal("expected update not to finish")
+		case <-time.After(time.Millisecond):
+		}
+		close(unblock)
+		require.NoError(t, <-updateErr)
+	}
+	for _, tc := range []struct {
+		name string
+		f    updateFunc
+	}{
+		{"started", (*jobs.Job).Started},
+		{"reverted", func(j *jobs.Job, ctx context.Context) error {
+			return j.Reverted(ctx, errors.New("boom"))
+		}},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			run(t, tc.f)
+		})
+	}
 }

